R/qenv-eval_code.R

```diff
@@ -28,18 +28,20 @@
 setGeneric("eval_code", function(object, code) standardGeneric("eval_code"))
 
 setMethod("eval_code", signature = c("qenv", "character"), function(object, code) {
-  id <- sample.int(.Machine$integer.max, size = 1)
+  parsed_code <- parse(text = code, keep.source = TRUE)
+  comments <- extract_comments(parsed_code)
+  id <- sample.int(.Machine$integer.max, size = length(parsed_code))
 
   object@id <- c(object@id, id)
   object@env <- rlang::env_clone(object@env, parent = parent.env(.GlobalEnv))
-  code <- paste(code, collapse = "\n")
-  object@code <- c(object@code, code)
+  object@code <- c(object@code, trimws(paste(as.character(parsed_code), comments)))
```

Comment:
comments should be extracted per single call. This makes `get_code` to return wrong code:

```r
pkgload::load_all("teal.code")

x <- qenv() |> eval_code("
  # initial comment
  a <- 1
  b <- 2 # inline comment
  c <- 3
  # inbetween comment
  d <- 4
  # finishing comment 
")

get_code(x)
# [1] "a <- 1 # initial comment"   "b <- 2 # inline comment"   
# [3] "c <- 3"                     "d <- 4 # inbetween comment"
# [5] "a <- 1 # finishing comment"
```

Reply:
@gogonzo thanks for finding that! It looks like this was caused by the inability of `utils::getParsedCode` to extract comments properly. We had `fix_shifted_comments` function that was used in `get_code_dependency`. I reused that function in `eval_code` to match comments with the calls. Thanks! also added 2 new tests

38bd16fb25b6a3a27d473d41e5694155cb29ccbf

```r
  # If comment is on top, it gets moved to the first call.
  # Any other comment gets moved to the call above.
  # Comments get pasted if there are two assigned to the same call.
library(teal.code)
code <- "
    # initial comment
    a <- 1 # A comment
    b <- 2 # inline comment
    c <- 3 # C comment
    # inbetween comment
    d <- 4
    # finishing comment
  "

qenv() |> eval_code(code) |> get_code()
#> [1] "a <- 1 # initial comment # A comment"  
#> [2] "b <- 2 # inline comment"               
#> [3] "c <- 3 # C comment # inbetween comment"
#> [4] "d <- 4 # finishing comment"
```

<sup>Created on 2024-10-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
Based on our conversation I prepared a (dirty prototype) function that divides the input code by calls and keeps comments as they were in the initial character input. 56b5c8e824383a4762c53f0dc78b566888eabeda

We need to decide if we want to pull the initial comments to the first call, and the final comments to the last call, as they are currently extracted as individual calls. Also the inline comment gets taken into the previous calls, not to the next call. This is also something that we can discuss.


```r
code <- "
  # initial comment line 1
  # initial comment line 2
  a <- 1 # A comment
  b1 <- 2; b2 <- 2;b3 = 3 # inline comment
  c <- 3 # C comment
  # inbetween comment
  d <- 4
  # finishing comment line 1
  # finishing comment line 2
"

parsed_code <- parse(text = code)
comments <- extract_comments(parsed_code)
pd <- utils::getParseData(parsed_code)

pd <- pd[pd$token != "';'", ]
lines_ids <- get_line_ids(pd)
code_by_calls <- split_code(code, lines_ids)
code_by_calls
```


```r
[[1]]
[1] "# initial comment line 1\n  # initial comment line 2"

[[2]]
[1] "a <- 1 # A comment"

[[3]]
[1] "b1 <- 2"

[[4]]
[1] "b2 <- 2"

[[5]]
[1] "b3 = 3 # inline comment"

[[6]]
[1] "c <- 3 # C comment\n  # inbetween comment"

[[7]]
[1] "d <- 4"

[[8]]
[1] "# finishing comment line 1\n  # finishing comment line 2"
```